File: src/signaling/signaling_client.cpp
// ============================================================
// 信令客户端 - 实现
// ============================================================

#include "signaling/signaling_client.h"
#include "common/protocol.h"
#include "common/utils.h"
#include <iostream>
#include <cstring>

// ============================================================
// 构造与析构
// ============================================================

SignalingClient::SignalingClient() {
}

SignalingClient::~SignalingClient() {
}

/**
 * @brief TCP探活 + DEVICE_HELLO 握手 (双方互相发现)
 */
bool SignalingClient::test_connect(const std::string& target_ip,
                                     uint16_t target_port,
                                     const std::string& my_id,
                                     const std::string& my_name,
                                     const std::string& my_ip,
                                     uint16_t my_port,
                                     uint32_t timeout_ms) {
    SOCKET_FD sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock == INVALID_SOCKET_FD) return false;

    NetworkUtils::set_nonblocking(sock);

    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons(target_port);
    if (inet_pton(AF_INET, target_ip.c_str(), &addr.sin_addr) <= 0) {
        CLOSE_SOCKET(sock);
        return false;
    }

    int conn_result = connect(sock, (struct sockaddr*)&addr, sizeof(addr));
    if (conn_result < 0) {
        int err = GET_SOCKET_ERROR();
        if (!NetworkUtils::is_would_block(err)) {
            CLOSE_SOCKET(sock);
            return false;
        }
    } else {
        // 立即连接成功, 发送 DEVICE_HELLO
        NetworkUtils::set_blocking(sock);
        json hello = Protocol::build_device_hello(my_id, my_name, my_ip, my_port);
        Protocol::send_json_message(sock, hello);
        // 优雅关闭确保数据被对方收到
        shutdown(sock, SHUT_WR);
        char dummy[64];
        while (recv(sock, dummy, sizeof(dummy), 0) > 0) {}
        CLOSE_SOCKET(sock);
        return true;
    }

    fd_set write_fds;
    FD_ZERO(&write_fds);
    FD_SET(sock, &write_fds);

    struct timeval timeout;
    timeout.tv_sec  = timeout_ms / 1000;
    timeout.tv_usec = (timeout_ms % 1000) * 1000;

    int select_result = select((int)(sock + 1), nullptr, &write_fds, nullptr, &timeout);

    if (select_result <= 0) {
        CLOSE_SOCKET(sock);
        return false;
    }

    int so_error = 0;
    socklen_t so_len = sizeof(so_error);
    bool ok = (getsockopt(sock, SOL_SOCKET, SO_ERROR, (char*)&so_error, &so_len) == 0 && so_error == 0);

    if (ok) {
        // 连接成功, 发送 DEVICE_HELLO 让对方发现本机
        NetworkUtils::set_blocking(sock);
        json hello = Protocol::build_device_hello(my_id, my_name, my_ip, my_port);
        Protocol::send_json_message(sock, hello);
        // 优雅关闭确保数据被对方收到
        shutdown(sock, SHUT_WR);
        char dummy[64];
        while (recv(sock, dummy, sizeof(dummy), 0) > 0) {}
    }

    CLOSE_SOCKET(sock);
    return ok;
}

// ============================================================
// 内部方法
// ============================================================

/**
 * @brief 向目标设备发送信令请求并接收响应
 *
 * 短连接模式:
 * 1. 连接到目标设备的信令端口
 * 2. 发送请求消息
 * 3. 接收响应消息
 * 4. 断开连接
 *
 * @param target_ip   目标设备IP地址
 * @param target_port 目标设备信令端口
 * @param request     请求消息
 * @param response    输出: 响应消息
 * @param timeout_ms  超时时间 (毫秒)
 * @return 是否成功
 */
bool SignalingClient::send_request(const std::string& target_ip,
                                    uint16_t target_port,
                                    const json& request,
                                    json& response,
                                    uint32_t timeout_ms) {
    // 1. 连接到目标设备
    SOCKET_FD sock = connect_to(target_ip, target_port, timeout_ms);
    if (sock == INVALID_SOCKET_FD) {
        std::cerr << "[信令客户端] 连接失败: " << target_ip << ":" << target_port << std::endl;
        return false;
    }

    // 2. 发送请求消息
    if (!Protocol::send_json_message(sock, request)) {
        std::cerr << "[信令客户端] 发送请求失败" << std::endl;
        CLOSE_SOCKET(sock);
        return false;
    }

    std::cout << "[信令客户端] 请求已发送: "
              << request.value("type", "") << " -> " << target_ip << ":" << target_port
              << std::endl;

    // 3. 接收响应消息
    if (!Protocol::recv_json_message(sock, response)) {
        std::cerr << "[信令客户端] 接收响应失败" << std::endl;
        CLOSE_SOCKET(sock);
        return false;
    }

    std::cout << "[信令客户端] 收到响应: "
              << response.value("type", "") << " ("
              << response.value("status", "N/A") << ")" << std::endl;

    // 4. 关闭连接
    CLOSE_SOCKET(sock);
    return true;
}

// ============================================================
// 内部方法: 带超时的TCP连接
// ============================================================

/**
 * @brief 创建TCP Socket并连接到目标设备
 *
 * 超时机制说明:
 * 标准的 connect() 默认是阻塞的, Linux上默认超时时间约 3 分钟,
 * 我们需要缩短这个超时时间来提升用户体验。
 *
 * 实现方法: 使用非阻塞Socket + select()
 * 1. 创建Socket, 设置为非阻塞
 * 2. 调用 connect() (非阻塞模式会立即返回, errno=EINPROGRESS)
 * 3. 使用 select() 或 poll() 等待指定的超时时间
 * 4. 连接成功后恢复为阻塞模式 (方便后续同步收发)
 */
SOCKET_FD SignalingClient::connect_to(const std::string& target_ip,
                                       uint16_t target_port,
                                       uint32_t timeout_ms) {
    // 1. 创建TCP Socket
    SOCKET_FD sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock == INVALID_SOCKET_FD) {
        std::cerr << "[信令客户端] socket() 失败: "
                  << NetworkUtils::get_last_error_string() << std::endl;
        return INVALID_SOCKET_FD;
    }

    // 2. 设置为非阻塞模式
    if (!NetworkUtils::set_nonblocking(sock)) {
        std::cerr << "[信令客户端] 设置非阻塞失败" << std::endl;
        CLOSE_SOCKET(sock);
        return INVALID_SOCKET_FD;
    }

    // 3. 构造目标地址
    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons(target_port);
    if (inet_pton(AF_INET, target_ip.c_str(), &addr.sin_addr) <= 0) {
        std::cerr << "[信令客户端] 无效的IP地址: " << target_ip << std::endl;
        CLOSE_SOCKET(sock);
        return INVALID_SOCKET_FD;
    }

    // 4. 发起连接 (非阻塞)
    // 非阻塞 connect() 通常会返回 -1 且 errno = EINPROGRESS
    // 表示连接正在建立中 (这是正常行为!)
    int conn_result = connect(sock, (struct sockaddr*)&addr, sizeof(addr));
    if (conn_result < 0) {
        int err = GET_SOCKET_ERROR();
        if (!NetworkUtils::is_would_block(err)) {
            // 不是 EINPROGRESS, 而是真正的错误
            std::cerr << "[信令客户端] connect() 错误: "
                      << NetworkUtils::get_error_string(err) << std::endl;
            CLOSE_SOCKET(sock);
            return INVALID_SOCKET_FD;
        }
    } else {
        // 连接立即成功 (罕见: 连接到本地主机时可能发生)
        NetworkUtils::set_blocking(sock);
        return sock;
    }

    // 5. 使用 select() 等待连接完成 (带超时)
    fd_set write_fds;
    FD_ZERO(&write_fds);
    FD_SET(sock, &write_fds);

    struct timeval timeout;
    timeout.tv_sec  = timeout_ms / 1000;
    timeout.tv_usec = (timeout_ms % 1000) * 1000;

    int select_result = select((int)(sock + 1), nullptr, &write_fds, nullptr, &timeout);

    if (select_result <= 0) {
        // 超时或错误
        if (select_result == 0) {
            std::cerr << "[信令客户端] 连接超时 (" << timeout_ms << "ms)" << std::endl;
        } else {
            std::cerr << "[信令客户端] select() 错误: "
                      << NetworkUtils::get_last_error_string() << std::endl;
        }
        CLOSE_SOCKET(sock);
        return INVALID_SOCKET_FD;
    }

    // 6. 检查连接是否真的成功
    // select 返回后可写, 但需要检查 SO_ERROR 确认连接是否成功
    int so_error = 0;
    socklen_t so_len = sizeof(so_error);
    if (getsockopt(sock, SOL_SOCKET, SO_ERROR,
                   (char*)&so_error, &so_len) < 0 || so_error != 0) {
        std::cerr << "[信令客户端] 连接失败: "
                  << NetworkUtils::get_error_string(so_error) << std::endl;
        CLOSE_SOCKET(sock);
        return INVALID_SOCKET_FD;
    }

    // 7. 恢复为阻塞模式 (方便后续同步收发)
    NetworkUtils::set_blocking(sock);

    return sock;
}

File: python/p2p_transfer/signaling/signaling_client.py
"""
TCP 信令客户端模块

提供两种使用模式：
1. send_request()：通用请求-应答模式，发送一条消息并等待响应
2. test_connect()：设备探测 + ECDH 密钥交换模式

非阻塞连接实现：
- connect_to() 使用经典的非阻塞 connect + select 超时模式
- 先设置 O_NONBLOCK → connect() → select() 等待 writable → 检查 SO_ERROR
- 连接成功后恢复为阻塞模式并设置超时

ECDH 握手流程（test_connect）：
1. 非阻塞 TCP 连接到目标设备的信令端口
2. 发送 DEVICE_HELLO（携带本机 ECDH 公钥）
3. 接收 DEVICE_HELLO_ACK（提取远端的 ECDH 公钥）
4. 计算共享密钥并存入 PeerKey
"""

import socket
import select
import errno
import time
from p2p_transfer.common.types import SIGNALING_PORT, MsgType
from p2p_transfer.common.platform import (
    set_nonblocking,
    set_blocking,
)
from p2p_transfer.common.protocol import (
    send_json_message,
    recv_json_message,
    build_device_hello,
)
from p2p_transfer.common.peer_key import PeerKey
from p2p_transfer.common.utils import compute_ecdh_shared


class SignalingClient:
    """短连接 TCP 信令客户端

    所有方法均为静态方法，无需实例化。
    """

    @staticmethod
    def connect_to(target_ip: str, target_port: int, timeout_ms: int = 5000) -> socket.socket | None:
        """非阻塞 TCP 连接到目标设备

        使用非阻塞 connect + select 超时模式：
        1. 创建 TCP 套接字
        2. 设置为非阻塞模式（O_NONBLOCK）
        3. 调用 connect()（预期返回 BlockingIOError 或 EINPROGRESS）
        4. 使用 select() 等待套接字变为可写
        5. 通过 getsockopt(SO_ERROR) 确认连接成功
        6. 恢复为阻塞模式并设置超时

        Args:
            target_ip: 目标 IP 地址
            target_port: 目标端口
            timeout_ms: 连接超时时间（毫秒）

        Returns:
            已连接的阻塞套接字，失败返回 None
        """
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        set_nonblocking(sock)

        # 非阻塞 connect，预期抛出 BlockingIOError（正常的 EINPROGRESS）
        try:
            sock.connect((target_ip, target_port))
        except BlockingIOError:
            pass                                    # 正常的连接进行中状态
        except OSError as e:
            sock.close()
            return None

        # 使用 select() 等待连接完成
        _, writable, _ = select.select([], [sock], [], timeout_ms / 1000.0)
        if not writable:
            sock.close()
            return None

        # 检查 SO_ERROR 确认连接成功
        err = sock.getsockopt(socket.SOL_SOCKET, socket.SO_ERROR)
        if err != 0:
            sock.close()
            return None

        # 连接成功，恢复阻塞模式
        set_blocking(sock)
        sock.settimeout(timeout_ms / 1000.0)
        return sock

    @staticmethod
    def send_request(
        target_ip: str,
        target_port: int,
        request: dict,
        timeout_ms: int = 5000,
    ) -> dict | None:
        """发送一条请求消息并等待应答

        完整的短连接流程：connect → send → recv → close。
        send/recv 自动根据 PeerKey 决定是否加密。

        Args:
            target_ip: 目标 IP
            target_port: 目标端口
            request: 要发送的 JSON 协议消息
            timeout_ms: 超时毫秒

        Returns:
            解析后的 JSON 应答字典，失败返回 None
        """
        sock = SignalingClient.connect_to(target_ip, target_port, timeout_ms)
        if sock is None:
            return None
        try:
            send_json_message(sock, request, target_ip)      # 自动加密
            response = recv_json_message(sock, target_ip)     # 自动解密
            return response
        except Exception as e:
            return None
        finally:
            try:
                sock.close()
            except:
                pass

    @staticmethod
    def test_connect(
        target_ip: str,
        target_port: int,
        device_id: str,
        device_name: str,
        local_ip: str,
        public_key: str,
        private_key: str,
        timeout_ms: int = 5000,
    ) -> bool:
        """设备探测连接 + ECDH 密钥交换

        此方法由主程序的探测线程调用，用于定期检查设备在线状态并完成 ECDH 握手。

        流程：
        1. 连接到目标设备信令端口
        2. 发送 DEVICE_HELLO（携带本机 ECDH 公钥）
        3. 接收 DEVICE_HELLO_ACK
        4. 提取远端的 ECDH 公钥
        5. 使用本地私钥和远端公钥计算共享密钥
        6. 将 32 字节共享密钥存入 PeerKey

        Args:
            target_ip: 目标设备 IP
            target_port: 目标信令端口（通常 8889）
            device_id: 本机 UUID
            device_name: 本机名称
            local_ip: 本机 IP
            public_key: 本机 ECDH 公钥（Base64）
            private_key: 本机 ECDH 私钥（Base64）
            timeout_ms: 超时毫秒

        Returns:
            握手是否成功
        """
        sock = SignalingClient.connect_to(target_ip, target_port, timeout_ms)
        if sock is None:
            return False

        try:
            sock.settimeout(timeout_ms / 1000.0)

            # 发送 DEVICE_HELLO 探测消息（携带本机 ECDH 公钥）
            hello = build_device_hello(
                device_id, device_name, local_ip, target_port, public_key
            )
            send_json_message(sock, hello, target_ip)

            # 接收 DEVICE_HELLO_ACK 应答
            response = recv_json_message(sock, target_ip)
            if response is None:
                return False

            msg_type = response.get("type", "")
            if msg_type != MsgType.DEVICE_HELLO_ACK:
                return False

            # 提取远端公钥并计算 ECDH 共享密钥
            remote_public = response.get("public_key", "")
            if remote_public:
                shared = compute_ecdh_shared(private_key, remote_public)
                if shared and len(shared) == 32:
                    # 将 32 字节共享密钥存入 PeerKey，后续通信自动加密
                    PeerKey.store(target_ip, shared)
                    return True

            return True

        except Exception:
            return False
        finally:
            try:
                sock.close()
            except:
                pass
